### PR TITLE
docs(policy)_: Reintroduced the testing policy for full review

### PR DESCRIPTION
**Pull request by Samyoul**
# PR Description: Reintroducing the `status-go` Testing Policy

## Summary
This PR reintroduces the `status-go` Testing Policy, providing clear guidelines for test creation, validation, and handling flaky tests. The policy has been updated to align with [Policy Zero](https://github.com/status-im/status-go/blob/develop/_docs/policies/introduce-policy.md), ensuring a structured, transparent, and inclusive approach to test management in the repository.  

## Background
The `status-go` Testing Policy previously existed in the repository but was removed to allow all Core Contributors (CCs) to review, discuss, and provide input before formal reintroduction. This aligns with Policy Zero, which establishes a framework for collaboratively developing policies in key areas of `status-go` development. The goal is to ensure that all policies reflect the needs and perspectives of contributors across the project.  

## What's Changed?*
- Reintroduces clear guidelines for creating tests, ensuring they are falsifiable, reliable, and resistant to fuzzing.  
- Defines flaky tests and mandates their resolution before merging.  
- Introduces structured steps for identifying, resolving, and reporting flaky tests.  
- Ensures compliance with Policy Zero’s submission, review, and approval processes.  
- Includes mermaid flowcharts for visual guidance on test resolution and reporting.  

## Why Change?  
- Establishes a consistent testing framework for `status-go`.  
- Prevents flaky or unreliable tests from being merged into the codebase.  
- Ensures that testing policies are maintained in an inclusive and transparent manner, following Policy Zero’s principles.  
- Gives all Core Contributors (CCs) an opportunity to engage with and shape this policy.

### CODE REVIEW

File: _docs/policies/testing.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+- Any individual MAY propose amendments to this policy.
+- Proposed changes MUST be submitted as a pull request (PR) to the `_docs/policies` directory in the `status-go` repository.
+- The PR MUST include a brief justification for the amendment, answering: "Why is this change needed?"
+- The PR MUST follow the [Review and Approval Process](./introduce-policy.md#review-and-approval-process) outlined in Policy Zero.
+
+## Creating Tests
+
+- All new functionality MUST be introduced with tests that:
+  - Prove that the functionality performs as described
+  - Can be falsified
```
Comment (friofry) [inline _docs/policies/testing.md:19]: > Can be falsified

I think it can be elaborated. Maybe it's just me, but I think these 2 possible interpretations are different:

* Tests SHOULD not only verify the correct ("happy path"). Edge-case and failure-scenario testing proactively validates the new feature resilience to unusual or invalid input

* Falsifiable testing retroactively ensures that a bug is demonstrably captured by a test, ensuring that fixes address concrete, reproducible problems rather than hypothetical scenarios.

**Comment by Samyoul** [comment]
Hey, as we refine this Testing Policy, I wanted to raise some questions for further discussion. Let me know what you think?

## API level Functional testing
Should API-level functional tests be explicitly included in this policy?
Right now, the policy covers functional testing in general, requiring tests to be falsifiable, reliable, and resistant to fuzzing, but it doesn’t explicitly differentiate API-level functional tests from other types of functional tests (e.g., integration or system tests).

Should API-level functional tests have their own specific requirements in this policy?
Are there any different considerations for API-level tests compared to other functional tests?
Do we need additional tooling, automation, or CI support to ensure API tests are properly covered in this policy?

### Performance & Load Testing
Should we introduce any policy items about performance benchmarks for functional tests, particularly API tests?
Should functional tests measure response times and resource consumption, or should that be handled separately in a `benchmarking policy` for example?

### Handling Test Debt & Maintenance
Do we need a process for reviewing and cleaning up outdated tests? 😬 
How do we prevent an accumulation of unreliable or useless tests over time?

### Fuzzing & Security Considerations
This policy mentions resistance to fuzzing, but should we also consider security testing within our tests?
Are we able to mock threats and attacks that are automatable via testing? 
Are there specific tooling or best practices we should standardise for security resilience?

### Automation & CI Integration
Should we mention in this policy the CI/CD automation steps currently in place that enforce test coverage?
Should we define requirements for test parallelisation, isolation, or dependencies?

### Flaky Test Management Process
Should we set a maximum timeframe for how long a test can remain flaky before it must be resolved or rewritten?

File: _docs/policies/testing.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+This policy defines the requirements and guidelines for testing within the `status-go` repository. It aims to ensure all new functionality is introduced with robust, reliable, and repeatable tests 
+while preventing flaky tests from being merged into the codebase. This aligns with Policy Zero’s principles of transparency, collaboration, and quality assurance in the development process.
+
+## Submitting and Maintaining the Test Policy
+
+- Any individual MAY propose amendments to this policy.
+- Proposed changes MUST be submitted as a pull request (PR) to the `_docs/policies` directory in the `status-go` repository.
+- The PR MUST include a brief justification for the amendment, answering: "Why is this change needed?"
+- The PR MUST follow the [Review and Approval Process](./introduce-policy.md#review-and-approval-process) outlined in Policy Zero.
```
Comment (igor-sirotin) [inline _docs/policies/testing.md:13]: I think this is redundant, as this is defined in the Policy Zero.

File: _docs/policies/testing.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+## Creating Tests
+
+- All new functionality MUST be introduced with tests that:
+  - Prove that the functionality performs as described
+  - Can be falsified
+  - Are resistant to fuzzing
```
Comment (igor-sirotin) [inline _docs/policies/testing.md:20]: Actually I don't think we need to specify the quality of the tests. I find our developers reasonable enough to implement tests that makes sense and cover the feature. This basically means "be a good developer, write good tests" 😄 We are good developers here!

Comment (igor-sirotin) [reply _docs/policies/testing.md:20]: So the point is: IMO just specifying that "the new functionality MUST be covered  with tests" is enough here.

File: _docs/policies/testing.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+- All new `functional tests` MUST BE validated via a minimum of 1000 test runs.
```
Comment (jrainville) [inline _docs/policies/testing.md:21]: I don't know functional tests that well, but `integration tests` are the usual culprits of flakiness, so either this is a swap or we need to add `integration` there too

**Review by jrainville — COMMENTED** [review]
Nice to see this back

File: _docs/policies/testing.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
+# `status-go` Test Policy
+
+## Purpose
+
+This policy defines the requirements and guidelines for testing within the `status-go` repository. It aims to ensure all new functionality is introduced with robust, reliable, and repeatable tests 
+while preventing flaky tests from being merged into the codebase. This aligns with Policy Zero’s principles of transparency, collaboration, and quality assurance in the development process.
```
Comment (ilmotta) [inline _docs/policies/testing.md:6]: > while preventing flaky tests from being merged into the codebase

In my experience, many flaky tests are tricky and end up in the main branch because they resist verification attempts. Beyond addressing flaky tests during PRs, an important process is to define how/who/when to handle flaky tests after they infect the main branch. But perhaps the details of such a process can be defined in a non-policy doc.

File: _docs/policies/testing.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+## Creating Tests
+
+- All new functionality MUST be introduced with tests that:
+  - Prove that the functionality performs as described
+  - Can be falsified
+  - Are resistant to fuzzing
+- All new `functional tests` MUST BE validated via a minimum of 1000 test runs.
+  - This can be achieved using the `-count` or `-test.count` flag with the test command, e.g., `-count 1000` / `-test.count 1000`.
+  - Where the CI cannot support this workflow automatically, the developer MUST perform validation tests via local testing.
```
Comment (ilmotta) [inline _docs/policies/testing.md:23]: How will the PR author prove the N number of test runs were executed? Should the CI be responsible for detecting _new_ functional tests and increase their test count by default for the duration of the PR? Is it always feasible to mandate a minimum of 1000 test runs when some tests can take a significant amount of time to finish?

File: _docs/policies/testing.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    - `TODO` Add link to issue for CI automation of validation test runs of new `functional tests`.
+  - Ensuring that the test passes consistently every time gives confidence that the test is not flaky.
```
Comment (ilmotta) [inline _docs/policies/testing.md:25]: We know one source of flakyness is external network dependency. I wonder if we can just define in status-go that certain tests (e.g. within a certain standardized directory like `./unit/` or following a specific naming pattern) should automatically block outbound traffic. I wonder if a hard block using a firewall in the CI could be practical, but it would be better I think if we could do from Go directly.

I'm just looking for ways to rely less on conventions and trust.

File: _docs/policies/testing.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    ]
+    F --> A
+    E --> G{Does the test appear in `E:Flaky Test` issues<br/> or in the last three nightly test runs?<br/>}
+    G -->|Yes| I[The flakiness needs reporting]
+    G -->|No| F
+    I --> J([Proceed to Reporting flow])
+```
+
+#### Reporting Flaky Tests
+If an old test fails and/or seems flaky either locally or in CI, you MUST report the event.
```
Comment (ilmotta) [inline _docs/policies/testing.md:69]: Is the lack of a blank line after the headline intentional? The other one at level 2 `## Flaky Tests` above has a blank line.

File: _docs/policies/testing.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+- Amendments:
+  - This policy MAY be amended at any time.
+  - Amendments MUST be submitted via a PR to the `status-go` repository.
+- Archival:
+  - This policy MAY be archived if it becomes obsolete or replaced by newer policies.
+  - Archival MUST be performed by submitting a PR that moves the policy to `_docs/policies/archived`.
+- The PR MUST include a rationale for the proposed amendment or archival in the PR description.
+- The PR MUST follow the [Review and Approval Process](./introduce-policy.md#review-and-approval-process) outlined in Policy Zero.
+
```
Comment (ilmotta) [inline _docs/policies/testing.md:120]: Two extra lines.

File: _docs/policies/testing.md (hunk 5 above)
Comment (ilmotta) [inline _docs/policies/testing.md:6]: Most lines don't have spurious line breaks, but some do like this one. For consistency with the other policy `introduce-policy.md`, it's better to fill all paragraphs at ~80 chars.

File: _docs/policies/testing.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+## Submitting and Maintaining the Test Policy
+
+- Any individual MAY propose amendments to this policy.
+- Proposed changes MUST be submitted as a pull request (PR) to the `_docs/policies` directory in the `status-go` repository.
+- The PR MUST include a brief justification for the amendment, answering: "Why is this change needed?"
+- The PR MUST follow the [Review and Approval Process](./introduce-policy.md#review-and-approval-process) outlined in Policy Zero.
+
+## Creating Tests
+
+- All new functionality MUST be introduced with tests that:
```
Comment (ilmotta) [inline _docs/policies/testing.md:17]: When it comes to the quality of tests, there are other characteristics not covered in the policy and we could go in more detail, but then, by going in more detail, I think we are already failing to use policies adequately. The policies are not great documents to iterate in my opinion and are too rigid and require too much coordination.

A few other characteristics:

- How sociable is the test?
- How does the test perform?
- How is the test structured in terms of arrange/act/assert?
- How much coverage should a test provide at a minimum?
- Is the test written using the best test pyramid level? Example: could the integration test be rewritten to be one or more unit tests, or a combination of different levels?
- Is the test verifying possible security threats?

> - This can be achieved using the `-count` or `-test.count` flag with the test command, e.g., `-count 1000` / `-test.count 1000`.

An example (to me) of something that seems too detailed to live in the policy.

**Review by ilmotta — COMMENTED** [review]
This is a welcome policy!

My impression while reading it is that it contains a bit too much detail, which leaves room for more amendments, something we mentioned in the introductory policy that we'd like to minimize.

For example, I don't think we need the policy to describe in detail how to resolve or report flaky tests (like in section `Steps to Resolving or Reporting Flaky Tests`). Another approach could be to link to external guidelines that can evolve independently of the policy. Essentially, I'm questioning what belongs in the policy versus what should live in non-policy docs.

File: _docs/policies/testing.md (hunk 4 above)
Comment (igor-sirotin) [reply _docs/policies/testing.md:21]: This should probably apply to unit/integration (go) tests, not functional (python). The latter have not yet been noted to be flaky, so I wouldn't apply such strict rules there.

We have tests types definition here: [status-go/Test Types](https://www.notion.so/Test-Types-3bbd74484f994491a11f4f1fe4a157af?pvs=4).

File: _docs/policies/testing.md (hunk 6 above)
Comment (igor-sirotin) [reply _docs/policies/testing.md:23]: I agree that a CI check would be nice, but the effort to implement it might not be worth the results.

There're 3 cases when a test can become flaky, and each reason requires different effort to implement a CI check:

| Reason | How to CI check | Implementation difficulty<br>(out of 5) | Warnings |
|-|-|-|-|
| When implementing a new test | Check if this test exists on `develop` | 🔴 🔴 | - | 
| When changing and old test code | Check tests diff and find amended tests | 🔴 🔴 🔴 🔴 | - |
| When changing tested code (without changing the test itself) | Use external tools (e.g. [Sentry Carryforward flags](https://about.codecov.io/blog/new-product-test-only-what-you-change-with-carryforward-flags/)) | 🔴 🔴 🔴 🔴 | Some code may be used in ALL tests, which would make CI to verify all tests a 1000 times | 

I think we could some day implement (1), 
but I would not bother with (2), 
and there will always be (3), which is wrong to do (note the warning above).

👉 So for now I believe it's fine to require devs to run new/amended test locally and perhaps attach the execution report to the PR description/comment. At the same time teach devs to prevent flaky tests, so that we won't even need to run tests a 1000 times.

File: _docs/policies/testing.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+  - Prove that the functionality performs as described
+  - Can be falsified
+  - Are resistant to fuzzing
+- All new `functional tests` MUST BE validated via a minimum of 1000 test runs.
+  - This can be achieved using the `-count` or `-test.count` flag with the test command, e.g., `-count 1000` / `-test.count 1000`.
```
Comment (igor-sirotin) [inline _docs/policies/testing.md:22]: I think this is unnecessary detail
```suggestion
  - This can be achieved using the `-count` or `-test.count` flag with the `go test` command
```

Comment (igor-sirotin) [reply _docs/policies/testing.md:22]: Btw, there's also `stress` tool, which is helpful: [status-go/Detect flaky tests](https://www.notion.so/Detect-flaky-tests-2ee4dba1f0e246a68981adc12fb6f652?pvs=4)

File: _docs/policies/testing.md (hunk 6 above)
Comment (igor-sirotin) [inline _docs/policies/testing.md:23]:
```suggestion
  - Where the CI cannot support this workflow automatically, the developer MUST perform validation tests locally.
```

File: _docs/policies/testing.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+## Policy Amendments and Archival
+
+- Amendments:
+  - This policy MAY be amended at any time.
+  - Amendments MUST be submitted via a PR to the `status-go` repository.
+- Archival:
+  - This policy MAY be archived if it becomes obsolete or replaced by newer policies.
+  - Archival MUST be performed by submitting a PR that moves the policy to `_docs/policies/archived`.
+- The PR MUST include a rationale for the proposed amendment or archival in the PR description.
+- The PR MUST follow the [Review and Approval Process](./introduce-policy.md#review-and-approval-process) outlined in Policy Zero.
```
Comment (igor-sirotin) [inline _docs/policies/testing.md:119]: This I believe is also redundant, as it is specified in the policy zero, which applies to all policies.

**Review by igor-sirotin — COMMENTED** [review]
Thank you @Samyoul! 

I believe it is essential for this policy to exist, and we should agree on common rules to introduce new tests, otherwise we will continue to introduce more and more flaky tests, and just keep skipping them when they become unbearably annoying.

P.S. This is just part of my review, I will come back with more comments on this. Just leaving this meanwhile.

File: _docs/policies/testing.md (hunk 6 above)
Comment (ilmotta) [reply _docs/policies/testing.md:23]: I had a feeling CI detection would be tricky, but I wasn't expecting such a detailed answer. Thanks @igor-sirotin 💯 I completely agree with your conclusions.

One idea to improve the effectiveness of the test suite is to randomly select a subset of tests and increase their run count (randomly as well if we like). Essentially, we'd extend the total test run time by a controlled amount. This would raise the chances of catching flaky tests before they hit `develop`. As the suite gets more stable, we could gradually reduce the randomness factor. We could also increase the number of affected tests by the random selection in CI by using dedicated jobs (e.g. twice per day with 10% of the test suite running between 10-200 times).

Is the idea above practical and perhaps straightforward?

**Comment by igor-sirotin** [comment]
Closing as policies are not used at the moment. Ref: https://github.com/status-im/status-go/pull/7112
